### ISSUE TITLE
SD3 T5 "CLIP" and PixArt interoperability

### ISSUE THREAD

**Issue by rsl8** [city96/ComfyUI_ExtraModels#61]
Interestingly, if T5 is the only encoder loaded as an "SD3-mode CLIP", the same conditionings can be used by both sd3 and pixart. Unfortunately it doesn't work, if "CLIP" incorporates all 3 text encoders.
I wonder if there is any way to use SD3 (with all 3 text encoders) and Pixart in the same workflow while avoiding loading T5 twice.

edit: T5, not T8. I don't know what is wrong with me :/

**Comment by city96** [comment]
Yeah, looks like it's doable. It needs an extra change to be 1:1 with the current implementation (SD3 uses padding, PixArt does not) but with that it's exactly the same as the current version. I'll add a `PixArtT5FromSD3` node tomorrow that splits T5 from the two CLIP text encoders to be used with PixArt.

![image](https://github.com/city96/ComfyUI_ExtraModels/assets/125218114/c99aac95-1eeb-4aab-9b7a-c4eba2813ea8)

**Comment by city96** [comment]
Done! Could you test it? I'll add it to the readme if it works correctly. I left the "padding" option in because it makes generating faster (cond/uncond pass can be batched) but changes the image a bit. A value of 1 (the default) should match the current version 1:1.

![image](https://github.com/city96/ComfyUI_ExtraModels/assets/125218114/ffa9ed40-17af-41b8-9933-a40254f9d02e)

**Comment by Extraltodeus** [comment]
You're looking too far! :D

![image](https://github.com/city96/ComfyUI_ExtraModels/assets/15731540/7d2ef7e0-20bf-4ef9-8469-82a30131e5e7)

Advanced > Loaders > Load CLIP

The simple one works, the result is the same whatever model you select in the bottom drop-down.

Workflow:

![00057UI_PixArt_00001_](https://github.com/city96/ComfyUI_ExtraModels/assets/15731540/16f67d67-3bd7-4876-adbb-fccb810efe54)

**Comment by city96** [comment]
@Extraltodeus Yeah, I did test with the simple one, and it does work out of the box minus the padding stuff.

The node I added adjusts the padding logic to the expected format so it's 1:1 for people that want to reproduce images made with this implementation (and I think it's closer to what the reference implementation does?).

It's also meant to work with the triple text encoder setup, which was the main problem this issue was about :P 

>Unfortunately it doesn't work, if "CLIP" incorporates all 3 text encoders.

>I wonder if there is any way to use SD3 (with all 3 text encoders) and Pixart in the same workflow while avoiding loading T5 twice.

**Comment by Extraltodeus** [comment]
I totally admit reading one word out of two lol.

**Comment by city96** [comment]
@Extraltodeus all good lol

**Comment by chrish-slingshot** [comment]
I've just been trying this and something very weird seems to be happening with the negative prompt. Just trying to put together a workflow to illustrate it. It looks like it's encoding the negative prompt differently to the default PixArt T5 way, but only for certain keywords? Bare with me.

**Comment by chrish-slingshot** [comment]
Looks like it's the way it handles weighting. Workflow in image attached.

![ComfyUI_temp_iklkr_00002_](https://github.com/city96/ComfyUI_ExtraModels/assets/117188274/dc118dd0-d27d-4968-b5bb-a1db8d8fa8b1)

**Comment by city96** [comment]
Oh yeah, weights are borked even without the node lol. Wonder why.

![image](https://github.com/city96/ComfyUI_ExtraModels/assets/125218114/ac894e49-3f10-4142-a75b-c2483435d25d)

**Comment by city96** [comment]
Looks like it's not as cut and dry as it seems. Based on just a quick test, PixArt prefers the extra models version, SD3 prefers the comfy version for how weights are handled.

My first guess would be that it's due to the weights using [no EOS token](https://github.com/city96/ComfyUI_ExtraModels/blob/main/T5/t5v11.py#L54) for my version and (presumably) using one for SD3 when calculating the weights.

![SDSDSD](https://github.com/city96/ComfyUI_ExtraModels/assets/125218114/dd816f42-131d-434e-b66b-11b013dfc7ff)

If anyone wants to test, adapting to SD3 seems to just involve doing this:

```py
class SD3EmbedFromPixArtEmbed:
	@classmethod
	def INPUT_TYPES(s):
		return {"required": {"pixart_emb": ("CONDITIONING",),}}

	RETURN_TYPES = ("CONDITIONING",)
	RETURN_NAMES = ("sd3_emb",)
	FUNCTION = "rep"
	TITLE = "SD3 T5 Embed from PixArt T5 Embed"

	def rep(self, pixart_emb):
		import copy
		emb = copy.deepcopy(pixart_emb)
		for k in emb:
			k[1]["pooled_output"] = torch.zeros([1, 2048])
		return (emb, )
```

**Comment by city96** [comment]
Sorry, adding the reference to the issue name in the commit closes it if you put "fix" in the title lol.

Though it *should* be fixed if anyone wants to retest it. The issue isn't with padding, it's the way the empty tokens are generated for the negative, as suspected above.

**Comment by chrish-slingshot** [comment]
All good now, nice one!

**Comment by city96** [comment]
@rsl8 Pretty sure the second deepcopy was left over from when I tried copying just the special tokens before realizing that the t5xxl object was still the same lol.

That trick with deepcopy looks neat, never seen that before. Still, I think I'll leave it for now considering it works, unless it somehow crashes in the middle it should be fine.
